### DISCUSSION TITLE
Usage of Vector API in Jlama

### DISCUSSION THREAD

**Discussion by anbusampath** [tjake/Jlama#19] [General]
I am new to LLMs, My understanding is that Java's Vector API used for SIMD instruction to auto-vectorization for different machine architecture. Where does SIMD used in Jlama? Because LLM uses embedding models, to communicate(input/output) with LLM why do we need to vectors.

**Comment by tjake** [comment]
Hi,

LLMs use many matrix multiplications.  In fact it's where 90% of the processing time goes when running inference.

You can run a matrix multiplication on a CPU with plain old java loops.  Or you can run them using Vector api to do it faster.  You can also run it on a GPU.

You can see two implementations in Jlama.

[Java Naive Operations](https://github.com/tjake/Jlama/blob/main/jlama-core/src/main/java/com/github/tjake/jlama/tensor/operations/NaiveTensorOperations.java)

[Java Panama Operations](https://github.com/tjake/Jlama/blob/main/jlama-core/src/main/java/com/github/tjake/jlama/tensor/operations/PanamaTensorOperations.java)

**Reply by anbusampath** [reply]
Thanks.
